repo2docker/app.py

```diff
@@ -741,7 +754,35 @@ def build(self):
                 shutil.rmtree(checkout_path, ignore_errors=True)
 
     def start(self):
-        self.build()
+        logfile = None
+        if (
+            self.s3_logs_endpoint
+            and self.s3_logs_access_key
+            and self.s3_logs_secret_key
+            and self.s3_logs_bucket
+        ):
+            logfile = tempfile.NamedTemporaryFile("w", delete=False)
+        try:
+            self.build(logfile=logfile)
+        finally:
+            if logfile:
+                logfile.close()
+                if os.stat(logfile.name).st_size:
+                    dest = f"buildlogs/{self.output_image_spec}/repo2docker.log"
+                    self.log.info(f"Uploading log to {self.s3_logs_bucket}/{dest}")
+                    s3 = boto3.resource(
+                        "s3",
+                        endpoint_url=self.s3_logs_endpoint,
+                        aws_access_key_id=self.s3_logs_access_key,
+                        aws_secret_access_key=self.s3_logs_secret_key,
+                        config=boto3.session.Config(signature_version="s3v4"),
+                        region_name=self.s3_logs_region,
+                    )
+                    s3.Bucket(self.s3_logs_bucket).upload_file(
+                        logfile.name,
```

Comment:
It's not quite as simple as uploading the terminal output record unmodified. With this recording, the file contains a *terminal stream* with ANSI escapes, carriage-returns, and other things that need processing. If you open them in a text editor without running through this, they will be very hard to read. If we want human-readable logs, we need to process the ANSI codes before uploading the resulting file.

Reply:
I found a regex from https://superuser.com/questions/380772/removing-ansi-color-codes-from-text-stream/380778#380778 that should match all ANSI escape sequences